### PR TITLE
SpaceConsistencyBear: Add blanklines_at_EOF option

### PR DESCRIPTION
**Pull request by khanchi97**
Add allow_trailing_blanklines option for whether to allow
trailing blanklines at EOF.

Closes https://github.com/coala/coala-bears/issues/1793

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [ ] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [ ] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: bears/general/SpaceConsistencyBear.py [hunk 1]
```diff
@@ -38,48 +41,84 @@ def run(self,
         result_texts = []
         additional_info_texts = []
 
+        def get_blanklines_nr_start():
+
```
Comment (Makman2) [inline bears/general/SpaceConsistencyBear.py]: though not violating PEP8, I usually prefer no blanklines between a `def` and code, this kinda interrupts reading flow imo. Optional though if you see it differently ;)

File: bears/general/SpaceConsistencyBear.py [hunk 2]
```diff
@@ -29,6 +30,8 @@ def run(self,
                                           of tabs.
         :param allow_trailing_whitespace: Whether to allow trailing whitespace
                                           or not.
+        :param allow_trailing_blanklines: Whether to allow trailing blanklines
+                                          at the EOF or not.
```
Comment (Makman2) [inline bears/general/SpaceConsistencyBear.py]: at ~~the~~ EOF

Comment (khanchi97) [reply bears/general/SpaceConsistencyBear.py]: okay

**Comment by Makman2** [comment]
The code looks way too complicated, I have the feeling you're touching too much old code. Is it possible to perform the check at the beginning, and then slice the lines in the file with `[x:y]` in `enumerate` in the `for`-loop?
So two cases:
1. don't allow trailing blanklines:
    analyze how many trailing blanklines are there, and suggest a patch. Modify `y` to make following analysis skip those lines.
2. allow trailing blanklines:
    Set `y` to end of file, and skip this analysis.

File: bears/general/SpaceConsistencyBear.py [hunk 3]
```diff
@@ -38,7 +41,52 @@ def run(self,
         result_texts = []
         additional_info_texts = []
 
-        for line_number, line in enumerate(file, start=1):
+        def get_blanklines_nr_start():
+            line_nr_start = False
+            if not allow_trailing_blanklines:
```
Comment (ishanSrt) [inline bears/general/SpaceConsistencyBear.py]: I think this should be `if allow_trailing_blanklines:` because only then you will have to check `if replacement.strip() == ''` otherwise `line_nr_start` [which I am guessing to be the *start from where blank lines or in case of trailing spaces allowed blank lines (with spaces) start]

File: bears/general/SpaceConsistencyBear.py [hunk 4]
```diff
@@ -38,7 +41,52 @@ def run(self,
...
+                rev_enumerated_zip_obj = zip(reversed(range(1, len(file) + 1)),
+                                             reversed(file))
+                rev_enumerated_tuple = tuple(rev_enumerated_zip_obj)
```
Comment (ishanSrt) [inline bears/general/SpaceConsistencyBear.py]: kind of defeats the purpose of having a `zip/generator` like object which is highly efficient than a `list like` object

File: bears/general/SpaceConsistencyBear.py [hunk 5]
```diff
@@ -38,7 +41,52 @@ def run(self,
...
+                for line_number, line in rev_enumerated_tuple:
+                    replacement = line
+                    if replacement.strip() == '':
+                        line_nr_start = line_number
+                    else:
+                        break
+
+            return line_nr_start
+
+        if allow_trailing_blanklines:
```
Comment (ishanSrt) [inline bears/general/SpaceConsistencyBear.py]: I think there should be a `not` in here. Please correct me if I am wrong

File: bears/general/SpaceConsistencyBear.py (hunk 3 above)
Comment (ishanSrt) [reply bears/general/SpaceConsistencyBear.py]: I stand a little bit corrected, if you `if` condition is true `line_nr_start` would be just the last line or False (in case there is no new line at EOF), if your `if` condition is incorrect then your `line_nr_start` would again be just the last line or False [in case there is no new line at EOF (even with spaces)] so don't quite get what you are trying to do over here

Comment (ishanSrt) [reply bears/general/SpaceConsistencyBear.py]: note there are 5 cases in which the file can end in case of trailing spaces are allowed

File: bears/general/SpaceConsistencyBear.py [hunk 6]
```diff
@@ -38,7 +41,52 @@ def run(self,
...
+                    'Line contains following spacing inconsistencies:'
+                    + inconsistencies,
+                    diffs={filename: diff},
+                    file=filename,
+                    line=end_line + 1,
+                    additional_info='\n\n'.join(additional_info_texts))
+                result_texts = []
+                additional_info_texts = []
+
+
```
[the remark this answers is not in the document]
Comment (ishanSrt) [reply bears/general/SpaceConsistencyBear.py]: this is not the end of a class so only 1 newline is expected here

File: bears/general/SpaceConsistencyBear.py (hunk 5 above)
Comment (khanchi97) [reply bears/general/SpaceConsistencyBear.py]: No.If trailing blanklines are allowed than then we don't to do any analysis for checking blanklines.

**Comment by khanchi97** [comment]
@ishanSrt Please don't review until I make required changes.It's not completed yet.I have to do some minor changes in bear and test as well.

File: bears/general/SpaceConsistencyBear.py [hunk 7]
```diff
@@ -38,7 +41,48 @@ def run(self,
         result_texts = []
         additional_info_texts = []
 
-        for line_number, line in enumerate(file, start=1):
+        def get_blanklines_nr_start():
+            line_nr_start = False
+            rev_enumerated_zip_obj = zip(reversed(range(1, len(file) + 1)),
+                                         reversed(file))
+            rev_enumerated_tuple = tuple(rev_enumerated_zip_obj)
```
Comment (ishanSrt) [inline bears/general/SpaceConsistencyBear.py]: Explain the reason for converting it into a tuple. You are losing the efficiency of a zip/generator like object by converting it into a list like object

**Comment by khanchi97** [comment]
>Then also your end_line_of_file is set to be the last line containing characters other that spaces.

No,it will set to the last non-blankline as `allow_trailing_blanklines` is not allowed.(acc. to you) ;)

**Comment by ishanSrt** [comment]
@khanchi97 
>allow_trailing_blanklines is not allowed

as in `allow_trailing_whitespace` and `allow_trailing_blankline` are both `False` and you have given a file which 
>Let us suppose there are lines in the end with only spaces.

and since to get val of `line_nr_start` you are stripping spaces and then checking, I'm afraid that 
>No,it will set to the last non-blankline

this may not hold true

File: bears/general/SpaceConsistencyBear.py (hunk 7 above)
Comment (khanchi97) [reply bears/general/SpaceConsistencyBear.py]: because zip return an iterator, so after zip return an iterator of tuples of `line and line no.` then to iterate over that iterator in `for loop` 👇 I converted it to a tuple.

Comment (khanchi97) [reply bears/general/SpaceConsistencyBear.py]: Can you a little bit explain how I'm losing the efficiency of zip by converting it to a tuple?

Comment (ishanSrt) [reply bears/general/SpaceConsistencyBear.py]: I think you can iterate normally in a for loop over an iterator, no need to convert it to tuple. Plus using iterators/generator like objects is highly efficient

Comment (ishanSrt) [reply bears/general/SpaceConsistencyBear.py]: >Can you a little bit explain how I'm losing the efficiency of zip by converting it to a tuple?

 list comprehension creates the entire list in memory first while the generator expression will create the items on the fly, so you are able to use it for very large (and also infinite!) sequences.

Comment (ishanSrt) [reply bears/general/SpaceConsistencyBear.py]: so if you only have to loop once then generator like objects are considered far more efficient and faster than list like objects. I think you can read the entire explanation on stackoverflow. there are many questions regarding this

Comment (khanchi97) [reply bears/general/SpaceConsistencyBear.py]: I think I've tried this when I was making changes but that didn't work.
btw, let me try this again.May be I was making some mistake ;)

**Comment by khanchi97** [comment]
oh, it will be set to `first_blankline_in_file` i.e `last_non-blankline +1`

File: bears/general/SpaceConsistencyBear.py (hunk 7 above)
Comment (khanchi97) [reply bears/general/SpaceConsistencyBear.py]: it working 👍 let see it won't error in CI

Comment (khanchi97) [reply bears/general/SpaceConsistencyBear.py]: thanks for your explanation :) @ishanSrt

File: .ci/package.json.jinja2 [hunk 8]
```diff
@@ -3,4 +3,3 @@
   "version": "{{version}}",
   "dependencies": {{dependencies | jsonify(indent=4, sort_keys=True)}}
 }
-
```
Comment (newbazz) [inline .ci/package.json.jinja2]: Why this change?

File: bears/general/SpaceConsistencyBear.py [hunk 9]
```diff
@@ -29,6 +30,8 @@ def run(self,
                                           of tabs.
         :param allow_trailing_whitespace: Whether to allow trailing whitespace
                                           or not.
+        :param allow_trailing_blanklines: Whether to allow trailing blanklines
+                                          at EOF or not.
```
Comment (newbazz) [inline bears/general/SpaceConsistencyBear.py]: Btw this is not the correct docstring indentation ;P

File: bears/general/SpaceConsistencyBear.py [hunk 10]
```diff
@@ -38,7 +41,47 @@ def run(self,
         result_texts = []
         additional_info_texts = []
 
-        for line_number, line in enumerate(file, start=1):
+        def get_blanklines_nr_start():
+            line_nr_start = False
```
Comment (newbazz) [inline bears/general/SpaceConsistencyBear.py]: variable names are not good ;P and can you please explain what exactly are you doing.

File: bears/general/SpaceConsistencyBear.py [hunk 11]
```diff
@@ -38,7 +41,47 @@ def run(self,
...
+            rev_enumerated_zip_obj = zip(reversed(range(1, len(file) + 1)),
+                                         reversed(file))
+            for line_number, line in rev_enumerated_zip_obj:
+                replacement = line
+                if replacement.strip() == '':
+                    line_nr_start = line_number
+                else:
+                    break
```
Comment (newbazz) [inline bears/general/SpaceConsistencyBear.py]: both `if` and `else` can be written inline.

File: .ci/package.json.jinja2 (hunk 8 above)
Comment (khanchi97) [reply .ci/package.json.jinja2]: this patch is applied by CI using this same bear.
I think it is applied because of the option that I've implemented i.e `allow_trailing_blanklines` which has default value `false`
@Makman2 ^ am i right?

File: bears/general/SpaceConsistencyBear.py (hunk 9 above)
Comment (khanchi97) [reply bears/general/SpaceConsistencyBear.py]: Well for now, it is 😉

File: bears/general/SpaceConsistencyBear.py (hunk 10 above)
Comment (khanchi97) [reply bears/general/SpaceConsistencyBear.py]: what do you suggest?

File: bears/general/SpaceConsistencyBear.py (hunk 11 above)
Comment (khanchi97) [reply bears/general/SpaceConsistencyBear.py]: I am not confirm we should do this or not?

File: bears/general/SpaceConsistencyBear.py (hunk 10 above)
Comment (ishanSrt) [reply bears/general/SpaceConsistencyBear.py]: what constitutes a good variable name?

**Comment by ishanSrt** [comment]
@khanchi97 I am not sure what you meant by https://github.com/coala/coala-bears/pull/2173#issuecomment-355223942 
Can you explain again why https://github.com/coala/coala-bears/pull/2173#issuecomment-355220629 is justified? This won't set to the last non blank line. As is said if the last lines only have spaces, it will completely ignore that and set it to the last line containing characters other than spaces. so the user won't even be prompted that he has trailing spaces in his file. Also since you are not evaluating that part of code, those lines will be left untouched and since the option of allow trailing spaces is false, and the final code will still contain trailing spaces, this is wrong output and you need to fix this. Again correct me if I am wrong. 🙂

File: bears/general/SpaceConsistencyBear.py (hunk 10 above)
Comment (khanchi97) [reply bears/general/SpaceConsistencyBear.py]: the name which is well suited for what it does and not so long and from it's name we can have a idea that what value it is holding(I guess 😉 )

**Comment by khanchi97** [comment]
@ishanSrt 
Is this type of file what you are asking about ?
```
file="int main(){\n"
      "a = 5;\n"
      "printf("%d",a);\n"
      "}\n"
      "     "
```

**Comment by ishanSrt** [comment]
more like
```
file="int main(){\n"
      "a = 5;\n"
      "printf("%d",a);\n"
      "}\n"
      "     \n"
      "\n"
```

**Comment by khanchi97** [comment]
okay so now as I have reversed all the file before start to check trailing blanklines.So now we will be checking the file in this order:
```
file=6.   "\n"
      5.    "     \n"
      4.    "}\n"
      3.    "printf("%d",a);\n"
      2.    "a = 5;\n"
      1.    "int main(){\n"
```
So now after exit from that `for loop` what do you think what would be the value of `line_nr_start`.
I think it should be 5.Now try to iterate this file into that `for loop` and tell me where do you have doubt?




      "\n"

**Comment by ishanSrt** [comment]
@khanchi97 Ok consider this file instead:
```
file="int main(){\n"
      "a = 5;\n"
      "printf("%d",a);\n"
      "}\n"
      "     \n"
      "   \n"
      "         \n"
      "\n"
```
so the reversed file is:
```
8      "\n"
7      "         \n"
6      "   \n"
5      "     \n"
4      "}\n"
3      "printf("%d",a);\n"
2      "a = 5;\n"
1      "int main(){\n"
```
now 
I think `line_nr_start` will return 5, so you are skipping lines 6 and 7

**Comment by khanchi97** [comment]
No.We are not skipping any line and it should give 5, shouldn't it?

**Comment by ishanSrt** [comment]
@khanchi97 so `end_line_of_file` would be set to **4** and your loop on line 84 will completely ignore lines 5,6, and 7 and thus the user won't be notified that there are trailing spaces

**Comment by khanchi97** [comment]
What do you think line 5,6,7 are not blanklines?

**Comment by ishanSrt** [comment]
won't the lines 5,6,7 would be left as it is in the file with no change? while the user selected the option to not allow trailing spaces and he is left with sort of trailing spaces in the final output file

**Comment by khanchi97** [comment]
No. the lines 5,6,7 would be deleted if `allow_trailing_blanklines` is FALSE and `allow_trailing_whitespace` is also FALSE.
And if `allow_traling_blanklines` is TRUE and `allow_trailing_whitespace` is FALSE then from lines 5,6,7 spaces will be removed.Fwew I think now you got it :)

**Comment by khanchi97** [comment]
And if you are not getting it then please once try to run it yourself and see what is the result :)

**Comment by ishanSrt** [comment]
@khanchi97 I stand a bit corrected. So the lines won't be added as it is as you have considered lines with spaces in the end to be `trailing newlines`. So the user isn't getting notified that he has trailing spaces. 

So, finally what should be done in this case? Shouldn't the user be notified that he has both trailing whitespaces and trailing newlines in the code rather than just removing those lines under trailing newlines. Not sure on this.

File: .ci/package.json.jinja2 (hunk 8 above)
Comment (newbazz) [reply .ci/package.json.jinja2]: I dont think so, coz CI needs to be configured in a way that no new file is being edited or made by it, and coala might have confgured CI correctly ;P. Still i am not sure.

File: bears/general/SpaceConsistencyBear.py (hunk 11 above)
Comment (newbazz) [reply bears/general/SpaceConsistencyBear.py]: does it violate any bear?

Comment (newbazz) [reply bears/general/SpaceConsistencyBear.py]: as far i remember when theres only one line in if statement it is good to inline it ;)

**Comment by newbazz** [comment]
ok thats what you have done ;P my bad

**Comment by khanchi97** [comment]
@ishan I think we should consider a line containing only a newline character "\n" and a line with whitespace also "   \n" to be a `blankline` and prompt the user that there are some trailing blanklines in your source code.

**Comment by khanchi97** [comment]
@Makman2 review needed 😄

File: .ci/package.json.jinja2 (hunk 8 above)
Comment (Makman2) [reply .ci/package.json.jinja2]: Yes it does immediately, due to the default value. Imo this is okay :+1: It's probably worth though to mention this inside the commit body :+1:

File: bears/general/SpaceConsistencyBear.py [hunk 12]
```diff
@@ -38,7 +41,47 @@ def run(self,
...
+
+            return line_nr_start
```
Comment (Makman2) [inline bears/general/SpaceConsistencyBear.py]: 1. You don't need this code to reside inside a function as you only use it once.
2. This can be simplified to a single expression:
    ```
    line_nr_start = next(i for i, line in reversed(list(enumerate(file))) if line.strip() != '')
    ```

File: bears/general/SpaceConsistencyBear.py [hunk 13]
```diff
@@ -38,7 +41,47 @@ def run(self,
...
+
+        if allow_trailing_blanklines:
+            end_line_of_file = len(file)
```
Comment (Makman2) [inline bears/general/SpaceConsistencyBear.py]: what about `blanklines_start_line`? I think that would be a good name :)
(yeah I finally got a name for it :D)

File: bears/general/SpaceConsistencyBear.py [hunk 14]
```diff
@@ -38,7 +41,47 @@ def run(self,
...
+
+        else:
+            blanklines_nr_start = get_blanklines_nr_start()
+            end_line_of_file = len(file)
+            if blanklines_nr_start:
```
Comment (Makman2) [inline bears/general/SpaceConsistencyBear.py]: Instead of letting the `get_blanklines_nr_start` function return `None` if no blanklines found, rather compare it against `len(file)`, makes things simpler :+1:

File: bears/general/SpaceConsistencyBear.py [hunk 15]
```diff
@@ -38,7 +41,47 @@ def run(self,
...
+                    break
+
+            return line_nr_start
+
+        if allow_trailing_blanklines:
+            end_line_of_file = len(file)
+
+        else:
+            blanklines_nr_start = get_blanklines_nr_start()
+            end_line_of_file = len(file)
```
Comment (Makman2) [inline bears/general/SpaceConsistencyBear.py]: seems this happens in both branches of the `if`

File: bears/general/SpaceConsistencyBear.py [hunk 16]
```diff
@@ -38,7 +41,47 @@ def run(self,
...
+                                          for string in result_texts)
+                yield Result.from_values(
+                    self,
+                    'Line contains following spacing inconsistencies:'
+                    + inconsistencies,
+                    diffs={filename: diff},
+                    file=filename,
+                    additional_info='\n\n'.join(additional_info_texts))
+                result_texts = []
+                additional_info_texts = []
```
Comment (Makman2) [inline bears/general/SpaceConsistencyBear.py]: See above, then you don't have to do this ;)

**Comment by Makman2** [comment]
unack 01f2a9dde9b5e5f6e7dc27825057f78a16c24a9c

**Comment by khanchi97** [comment]
@Makman2 thank you very much for your review 😄 
will do the required changes by tomorrow.
I was a bit busy in organizing #LinuxWeek for our juniors.
(going to sleep 😛 )

**Comment by khanchi97** [comment]
I have made the required changes :)

File: bears/general/SpaceConsistencyBear.py [hunk 17]
```diff
@@ -38,7 +41,36 @@ def run(self,
...
+        else:
+            blanklines_start_line = line_nr_start + 1 \
+                    if line_nr_start < len(file) else False
+
+        if not allow_trailing_blanklines:
+            if blanklines_start_line:
+                additional_info_text = (
+                        'Your source code contains trailing blanklines at EOF.'
+                        'Those usually have no meaning. Please consider'
+                        'removing them.')
```
[the remark this answers is not in the document]
Comment (Makman2) [reply bears/general/SpaceConsistencyBear.py]: Ah nvm that's the *additional*-info-text

Comment (Makman2) [inline bears/general/SpaceConsistencyBear.py]: Indentation: 4 spaces relative to statement

File: bears/general/SpaceConsistencyBear.py [hunk 18]
```diff
@@ -38,7 +41,36 @@ def run(self,
...
-        for line_number, line in enumerate(file, start=1):
+        line_nr_start = next((i for i, line in reversed(list(
+            enumerate(file, start=1))) if line.strip() != ''), 0)
+
+        if len(file) == 0:
+            blanklines_start_line = False
+
+        else:
+            blanklines_start_line = line_nr_start + 1 \
+                    if line_nr_start < len(file) else False
```
Comment (Makman2) [inline bears/general/SpaceConsistencyBear.py]: I have the feeling you are unnecessary trying to use special values (--> `False`). 

You can easily check if you have to yield a result if `line_nr_start < len(file)`.
The whole stuff can be refactored to roughly this:
```python
if not allow_trailing_blanklines:
    line_nr_start = next((i for i, line in reversed(list(
        enumerate(file, start=1))) if line.strip() != ''), 0)
    # TODO: Maybe more performance: Don't increment from start=1, but start=0 and
    #       the next-default value = -1, and increment in the end: next(..., -1) + 1

    if line_nr_start < len(file):
        # Note: You have to provide line_nr_start + 1 to refer to the right lines.
        # line_nr_start is right now designed to give the starting line referring
        # to zero-based indices.
        yield Result(...)
else:
    line_nr_start = len(file)

for line_number, line in enumerate(file[:line_nr_start])
```

Comment (Makman2) [reply bears/general/SpaceConsistencyBear.py]: Note also how I've put the relevant analysis inside the if, to save performance in case someone wants not to take care of blanklines

File: bears/general/SpaceConsistencyBear.py [hunk 19]
```diff
@@ -38,7 +41,36 @@ def run(self,
         result_texts = []
         additional_info_texts = []
 
-        for line_number, line in enumerate(file, start=1):
+        line_nr_start = next((i for i, line in reversed(list(
+            enumerate(file, start=1))) if line.strip() != ''), 0)
```
Comment (Makman2) [inline bears/general/SpaceConsistencyBear.py]: Optional:
Is it maybe possible to wrap the statement somehow so it's more readable?

Comment (Makman2) [reply bears/general/SpaceConsistencyBear.py]: Something in the direction of
```python
line_nr_start = next(
    (i
     for i, line in reversed(list(enumerate(file, start=1)))
     if line.strip() != ''), 0)
```

File: tests/general/SpaceConsistencyBearTest.py [hunk 20]
```diff
@@ -70,3 +75,22 @@ def test_enforce_newline_at_eof(self):
...
+                               "  print('Is it funny?')\n",
+                               "  print('Yeah, funny')\n",
+                               '  \n',
+                               '\n'],
+                              force_linebreaks=False)
+        self.check_validity(self.uut,
+                            ['def funny_code():\n',
+                             "  print('Is it funny?')\n",
+                             "  print('Yeah, funny')\n"],
+                            force_linebreaks=False)
```
Comment (Makman2) [inline tests/general/SpaceConsistencyBearTest.py]: I want to see tests for
- an empty file
- a file with only a single blank line
- a file with only multiple blank lines
- a file with leading and trailing blank lines

Comment (Makman2) [inline tests/general/SpaceConsistencyBearTest.py]: And please use `check_results` to assert for the bear results :+1:

**Comment by Makman2** [comment]
unack 2e2eaaceb73a96fae24e89919ca02bd828729109

File: bears/general/SpaceConsistencyBear.py (hunk 18 above)
Comment (khanchi97) [reply bears/general/SpaceConsistencyBear.py]: yeah, I wanted to do the same.But I was finding it difficult to cover the case when file is empty.That's why I used that (--> False), but now it will work as we are using `start=0` and next default value `-1`.

**Comment by Makman2** [comment]
please fix the gitmate issues ;)

**Comment by jayvdb** [comment]
Conflicts to be resolved.

**Comment by khanchi97** [comment]
@Makman2 @jayvdb Can I ask for adding someone as collaborator to my PRs(which are almost completed) to do remaining work. ( I am busy in some other work :) )

File: bears/general/SpaceConsistencyBear.py [hunk 21]
```diff
@@ -38,7 +41,34 @@ def run(self,
...
+                    'Trailing blanklines found.',
+                    diffs={filename: diff},
+                    file=filename,
+                    line=line_nr_start + 1,
+                    column=1,
+                    end_line=len(file),
+                    end_column=len(file[-1]),
+                    additional_info=additional_info_text)
+
+        else :
```
Comment (sangamcse) [inline bears/general/SpaceConsistencyBear.py]: Fix this whitespace issue @khanchi97. Run `coala` over your local repository before pushing it to remote 😉

File: bears/general/SpaceConsistencyBear.py [hunk 22]
```diff
@@ -18,6 +18,7 @@ def run(self,
             file,
             use_spaces: bool,
             allow_trailing_whitespace: bool=False,
+            allow_trailing_blanklines: bool=False,
```
Comment (sangamcse) [inline bears/general/SpaceConsistencyBear.py]: Alphabetical order 😉

File: tests/general/SpaceConsistencyBearTest.py [hunk 23]
```diff
@@ -44,23 +46,28 @@ def test_data_sets_tabs(self):
         self.section.append(Setting('use_spaces', 'false'))
         self.section.append(Setting('allow_trailing_whitespace', 'true'))
         self.section.append(Setting('enforce_newline_at_EOF', 'false'))
+        self.section.append(Setting('allow_trailing_blanklines', 'false'))
```
Comment (sangamcse) [inline tests/general/SpaceConsistencyBearTest.py]: This line is not required.

File: tests/general/SpaceConsistencyBearTest.py [hunk 24]
```diff
@@ -70,3 +77,35 @@ def test_enforce_newline_at_eof(self):
...
                               force_linebreaks=False)
+
+    def test_trailing_blanklines(self):
+
+        self.maxDiff = None
+        self.section['use_spaces'] = 'true'
+        self.section['allow_trailing_whitespace'] = 'false'
+        self.section['enforce_newline_at_EOF'] = 'true'
+        self.section['allow_trailing_blanklines'] = 'false'
+        self.check_results(self.uut,
```
Comment (sangamcse) [inline tests/general/SpaceConsistencyBearTest.py]: What do you say about this?
```
    def test_trailing_blanklines(self):
        self.maxDiff = None
        self.section.append(Setting('use_spaces', 'true'))
        self.section.append(Setting('allow_trailing_whitespace', 'false'))
        self.section.append(Setting('enforce_newline_at_EOF', 'true'))
        self.section.append(Setting('allow_trailing_blanklines', 'false'))
        self.check_validity(self.uut,
                            ['sangam\n'],
                            force_linebreaks=False)
        self.check_invalidity(self.uut,
                              [' \n'],
                              force_linebreaks=False)
```
